internal/proxymanager/proxymanager.go

```diff
@@ -0,0 +1,117 @@
+// Package proxymanager implements an API for configuring proxy-related settings
+// on the client machine.
+package proxymanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ubuntu/adsys/internal/decorate"
+	log "github.com/ubuntu/adsys/internal/grpc/logstreamer"
+	"github.com/ubuntu/adsys/internal/i18n"
+)
+
+type options struct {
+	aptConfigPath         string
+	environmentConfigPath string
+}
+type option func(*options)
+
+// Manager prevents multiple writes to the configuration files in parallel.
+type Manager struct {
+	proxies []proxySetting
+
+	aptConfigPath         string
+	environmentConfigPath string
+
+	applyMu sync.Mutex
+}
+
+// New returns a new instance of a proxy manager.
+func New(ctx context.Context, config Config, opts ...option) (m *Manager, err error) {
+	defer decorate.OnError(&err, i18n.G("couldn't create proxy manager"))
+
+	// defaults
+	args := options{
+		environmentConfigPath: "/etc/environment.d/99adsys-proxy.conf",
+		aptConfigPath:         "/etc/apt/apt.conf.d/99adsys-proxy",
+	}
+	// applied options
+	for _, o := range opts {
+		o(&args)
+	}
+
+	proxies, err := setConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Manager{
+		proxies:               proxies,
+		aptConfigPath:         args.aptConfigPath,
+		environmentConfigPath: args.environmentConfigPath,
+	}, nil
+}
+
+// Apply applies the proxy configuration to the system.
+func (m *Manager) Apply(ctx context.Context) error {
+	m.applyMu.Lock()
+	defer m.applyMu.Unlock()
+
+	if err := m.applyEnvironmentProxy(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Manager) applyEnvironmentProxy(ctx context.Context) (err error) {
+	defer decorate.OnError(&err, i18n.G("couldn't apply environment proxy configuration"))
+
+	content := "### This file was generated by ADSys - manual changes will be overwritten\n"
+	for _, p := range m.proxies {
+		content += p.envString()
+	}
+
+	if exists, prevContent, err := prevConfIfExists(m.environmentConfigPath); exists && prevContent == content {
+		log.Debugf(ctx, fmt.Sprintf("Environment proxy configuration at %q is already up to date", m.environmentConfigPath))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Check if the parent directory exists - attempt to create the structure if not
+	environmentConfigDir := filepath.Dir(m.environmentConfigPath)
+	if _, err := os.Stat(filepath.Dir(m.environmentConfigPath)); errors.Is(err, os.ErrNotExist) {
```

Comment:
You already declared the `environmentConfigDir` var above with the `filepath.Dir` returned value, there's no need to call it again inside `os.Stat`